File: python/src/analytics_kit/flags/local/definition_poller.py
"""The definition poller — fetch flag DEFINITIONS (not evaluated flags) on an interval, parse them
into an in-memory :class:`DefinitionSnapshot` the evaluator reads.

The ONLY blocking-I/O boundary in local eval lives here (the definitions fetch); the matcher is a
pure synchronous function. Blocking round-trip behind a background thread, exactly the posture the
E12 remote flag adapter and the HTTP query adapter use — no asyncio.

Poll-thread hygiene (the E12-S4 daemon-thread-leak lesson): the loop is a ``threading.Event``-gated
daemon whose :meth:`stop` SETS the event and JOINS the thread, so no thread leaks past a test that
stops it. Concurrent :meth:`load` calls dedup onto a single in-flight lock so only one fetch runs at
a time. An injectable transport (the same :class:`~analytics_kit.flags.transport.FlagTransport` seam
the remote adapter uses) so tests never hit a live backend.
"""

# De-branded from posthog's poller.py (the definition poll loop), feature_flags.py
# _load_feature_flags / load_feature_flags / update_flag_state, and flag_definition_cache.py.

from __future__ import annotations

import json
import threading

from ..transport import FlagTransport, _UrllibFlagTransport
from .definition_types import (
    EMPTY_SNAPSHOT,
    DefinitionSnapshot,
    FlagDefinition,
    PropertyGroup,
)

# --- Wire definition-fetch vocabulary ------------------------------------------------------------
#
# The definitions endpoint path, its query params, the auth scheme, and the response keys — all
# adapter-internal, confined to these module-level ``_WIRE_*`` constants and never leaving this
# module. A future backend adapter negotiating a different definition wire supplies its own. The
# neutral config carries only ``definitions_endpoint``/``definitions_key``/``token``, never these.
# De-branded from posthog's node definitions request + response.
_WIRE_DEFINITIONS_PATH = "/flags/definitions"
_WIRE_TOKEN_QUERY = "token"
_WIRE_SEND_COHORTS_QUERY = "send_cohorts"
_WIRE_FLAGS_KEY = "flags"
_WIRE_GROUP_TYPE_MAPPING_KEY = "group_type_mapping"
_WIRE_COHORTS_KEY = "cohorts"

_WIRE_METHOD_GET = "GET"
_WIRE_CONTENT_TYPE_HEADER = "Content-Type"
_WIRE_CONTENT_TYPE_JSON = "application/json"
_WIRE_AUTHORIZATION_HEADER = "Authorization"
_WIRE_BEARER_SCHEME = "Bearer"

_STATUS_OK_FLOOR = 200
_STATUS_OK_CEIL = 300


class DefinitionPoller:
    """Poll flag definitions on an interval into an in-memory snapshot the evaluator reads.

    ``definitions_key`` is the privileged (definition-reading) credential, named BY ROLE — never a
    vendor key name; it authorizes the definition fetch and is distinct from the ingest write key and
    the remote-eval project key. ``token`` scopes the definitions to a project. The transport is the
    injectable HTTP send hook (default a stdlib ``urllib`` GET).
    """

    def __init__(
        self,
        *,
        definitions_endpoint: str,
        definitions_key: str,
        token: str,
        poll_interval: float,
        transport: FlagTransport | None = None,
    ) -> None:
        self._url = _resolve_definitions_url(definitions_endpoint, token)
        self._definitions_key = definitions_key
        self._poll_interval = poll_interval
        self._transport: FlagTransport = transport if transport is not None else _UrllibFlagTransport()
        self._snapshot: DefinitionSnapshot = EMPTY_SNAPSHOT
        self._loaded_successfully_once = False
        # In-flight dedup: only one fetch runs at a time; a concurrent load waits on this lock and
        # sees the just-loaded snapshot instead of issuing a second request.
        self._load_lock = threading.Lock()
        self._stopped = threading.Event()
        # Set once the daemon thread's FIRST load attempt has completed (success OR failure). The
        # deterministic "the fire-and-forget first load has drained" signal — the sync analog of
        # awaiting the TS async `start()`. Production never blocks on it (is_ready() gates eval);
        # only a test that needs the first load settled before asserting readiness waits on it.
        self._first_load_done = threading.Event()
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        """Kick off polling and RETURN AT ONCE — the daemon thread does the first load as its own
        first step, so construction never blocks on the definitions round-trip (an unreachable
        endpoint must not hang app boot). ``is_ready()`` gates local eval, so ``evaluate`` falls
        through to remote/degraded until that first load lands. A background loop then reloads every
        ``poll_interval`` until :meth:`stop`. Idempotent — a second ``start`` is a no-op."""
        if self._thread is not None:
            return
        self._thread = threading.Thread(
            target=self._run, name="analytics-kit-flag-poller", daemon=True
        )
        self._thread.start()

    def is_ready(self) -> bool:
        """True once at least one successful load parsed a non-empty definition list — the "should I
        try local eval" gate."""
        return self._loaded_successfully_once and len(self._snapshot.flags) > 0

    def wait_for_first_load(self, timeout: float | None = None) -> bool:
        """Block until the daemon thread's FIRST load ATTEMPT has completed (success OR failure), then
        report whether the wait actually observed that completion (``False`` on timeout).

        The synchronous analog of awaiting the TS async ``start()``: since ``start()`` returns before
        the fire-and-forget first load runs, a caller that needs the load settled (a test asserting
        local eval, or the negative case) waits here. Completion is NOT the same as readiness — a
        failed first load completes but leaves :meth:`is_ready` ``False``; check ``is_ready()``
        separately after this returns (mirroring the TS ``await start(); expect(isReady())`` sequence).
        Never called on the production path (eval falls through to remote until the load lands)."""
        return self._first_load_done.wait(timeout)

    def get_snapshot(self) -> DefinitionSnapshot:
        """The current parsed definition snapshot, read atomically by the evaluator. Before the first
        successful load this is the frozen empty snapshot (never ``None``)."""
        return self._snapshot

    def stop(self) -> None:
        """Halt polling: SET the stop event and JOIN the poll thread so no thread leaks. Idempotent —
        this is the E12-S4 leak fix; every test that starts the poller calls it in teardown."""
        self._stopped.set()
        thread = self._thread
        if thread is not None and thread.is_alive():
            thread.join(timeout=5)
        self._thread = None

    def load(self) -> None:
        """Load definitions once, deduping a concurrent call onto the same in-flight lock. A failed
        load leaves the prior snapshot in place — never overwrite good data with an error; the next
        scheduled poll retries."""
        with self._load_lock:
            self._fetch_definitions()

    def _run(self) -> None:
        # The first load runs HERE, on the thread — never on the caller's stack — so construction
        # returns before any network I/O. The reference daemon posture is a bare `threading.Thread`
        # with `daemon=True` and no clean stop (the E12-S4 leak source); here the loop is
        # `Event`-gated: `wait(interval)` returns True the instant `stop()` sets the event, so the
        # loop exits promptly and `stop()`'s join returns — no leaked thread.
        try:
            self.load()
        finally:
            # Signal completion in a `finally`, UNCONDITIONALLY — an unexpected raise in the first
            # load must still release a `wait_for_first_load` waiter (else it hangs to its timeout).
            self._first_load_done.set()
        while not self._stopped.wait(self._poll_interval):
            self.load()

    def _fetch_definitions(self) -> None:
        try:
            response = self._transport.send(
                self._url,
                _WIRE_METHOD_GET,
                {
                    _WIRE_CONTENT_TYPE_HEADER: _WIRE_CONTENT_TYPE_JSON,
                    _WIRE_AUTHORIZATION_HEADER: f"{_WIRE_BEARER_SCHEME} {self._definitions_key}",
                },
            )
        except Exception:  # noqa: BLE001 — a transport failure leaves the prior snapshot; the next poll retries.
            return
        if not _STATUS_OK_FLOOR <= response.status < _STATUS_OK_CEIL:
            return
        parsed = _parse_definitions(response.body)
        if parsed is None:
            return
        self._snapshot = parsed
        self._loaded_successfully_once = True


def _parse_definitions(body: str) -> DefinitionSnapshot | None:
    """Parse the wire definitions response into a :class:`DefinitionSnapshot`. ``None`` on a malformed
    body — the caller leaves the prior snapshot in place."""
    try:
        decoded = json.loads(body)
    except (ValueError, TypeError):
        return None
    if not isinstance(decoded, dict):
        return None
    raw_flags = decoded.get(_WIRE_FLAGS_KEY)
    flags: tuple[FlagDefinition, ...] = tuple(raw_flags) if isinstance(raw_flags, list) else ()
    flags_by_key = {str(flag.get("key")): flag for flag in flags if isinstance(flag, dict)}
    group_type_mapping = decoded.get(_WIRE_GROUP_TYPE_MAPPING_KEY)
    cohorts = decoded.get(_WIRE_COHORTS_KEY)
    return DefinitionSnapshot(
        flags=flags,
        flags_by_key=flags_by_key,
        group_type_mapping=dict(group_type_mapping) if isinstance(group_type_mapping, dict) else {},
        cohorts=_cohorts(cohorts),
    )


def _cohorts(raw: object) -> dict[str, PropertyGroup]:
    return {str(k): v for k, v in raw.items()} if isinstance(raw, dict) else {}


def _resolve_definitions_url(definitions_endpoint: str, token: str) -> str:
    """Resolve the definitions URL from the consumer's bare flag origin: append the definitions path
    and the token + send-cohorts query params."""
    host = definitions_endpoint.strip().rstrip("/")
    return f"{host}{_WIRE_DEFINITIONS_PATH}?{_WIRE_TOKEN_QUERY}={token}&{_WIRE_SEND_COHORTS_QUERY}="

File: ts/packages/node/src/flags/local/definition-poller.ts
import type { FetchLike } from '../../config';
import { joinHostPath } from '../../ingest-url';
import type {
  DefinitionSnapshot,
  FlagDefinition,
  PropertyGroup,
} from './definition-types';

// The definition poller, de-branded from posthog's node FeatureFlagsPoller load/reschedule path.
// It fetches flag DEFINITIONS (not evaluated flags) on an interval, parses them into an in-memory
// DefinitionSnapshot the evaluator reads, and exposes `isReady()` + a snapshot accessor S2 binds
// to. The only async boundary in local eval lives here (the fetch); the matcher is synchronous.
// De-branded from posthog's feature-flags.ts _loadFeatureFlags / loadFeatureFlags / updateFlagState
// / _requestFeatureFlagDefinitions.

// Adapter-internal [WIRE] definition-fetch vocabulary: the endpoint path, its query params, and the
// response keys the definitions endpoint speaks. Plain UPPER_SNAKE `*_WIRE_*` consts — the SHIPPED
// node flag convention (non-`$`; the `$`-const style is the browser package's). None of it appears
// on the neutral surface. A future backend adapter negotiating a different definition wire supplies
// its own.

// The definition-list path appended to the consumer's flag-endpoint origin.
const DEFINITIONS_ENDPOINT_WIRE_PATH = '/flags/definitions';
// The query param carrying the project token, and the flag that asks the endpoint to include the
// cohort map in the response.
const TOKEN_WIRE_QUERY = 'token';
const SEND_COHORTS_WIRE_QUERY = 'send_cohorts';
// Response keys carrying the definition list, the group-type mapping, and the cohort map.
const FLAGS_WIRE_KEY = 'flags';
const GROUP_TYPE_MAPPING_WIRE_KEY = 'group_type_mapping';
const COHORTS_WIRE_KEY = 'cohorts';

// The [WIRE] definition-fetch response the endpoint returns. Adapter-internal — never neutral
// surface.
interface DefinitionsWireResponse {
  [FLAGS_WIRE_KEY]?: FlagDefinition[];
  [GROUP_TYPE_MAPPING_WIRE_KEY]?: Record<string, string>;
  [COHORTS_WIRE_KEY]?: Record<string, PropertyGroup>;
}

// The poller's adapter-internal config. `definitionsKey` is the privileged (definition-reading)
// credential, named BY ROLE — never a vendor key name; it authorizes the definition fetch and is
// distinct from the ingest write key and the remote-eval project key. `token` is the project token
// the endpoint scopes the definitions to.
export interface DefinitionPollerConfig {
  definitionsEndpoint: string;
  definitionsKey: string;
  token: string;
  pollIntervalMs: number;
  fetch: FetchLike;
}

const EMPTY_SNAPSHOT: DefinitionSnapshot = Object.freeze({
  flags: Object.freeze([]),
  flagsByKey: Object.freeze({}),
  groupTypeMapping: Object.freeze({}),
  cohorts: Object.freeze({}),
});

export class DefinitionPoller {
  // The fetch machinery is UNDEFINED in seeded mode — the seeded poller structurally lacks the URL,
  // the privileged credential, and the fetch it would need to reach a definitions endpoint (S2's
  // structural guardrail: "no fetch, no URL, no thread" is a property of construction, not merely of
  // the start() no-op). A fetching poller sets all three; a seeded one leaves them undefined and its
  // start()/load() short-circuit before ever touching them.
  private readonly definitionsUrl: string | undefined;
  private readonly definitionsKey: string | undefined;
  private readonly pollIntervalMs: number | undefined;
  private readonly doFetch: FetchLike | undefined;
  private snapshot: DefinitionSnapshot = EMPTY_SNAPSHOT;
  private loadedSuccessfullyOnce = false;
  private timer: ReturnType<typeof setTimeout> | undefined;
  // In-flight dedup: concurrent load requests share this single promise so only one fetch is in
  // flight at a time; cleared when it settles.
  private loadingPromise: Promise<void> | undefined;
  private stopped = false;
  // Seeded mode: definitions came from config (S2), not the fetch. `start()` is a real no-op and the
  // seed is already `isReady()` from construction — the adapter reads through the same `local.poller`
  // field and cannot tell the difference.
  private readonly seeded: boolean;

  constructor(config: DefinitionPollerConfig) {
    this.definitionsUrl = resolveDefinitionsUrl(config.definitionsEndpoint, config.token);
    this.definitionsKey = config.definitionsKey;
    this.pollIntervalMs = config.pollIntervalMs;
    this.doFetch = config.fetch;
    this.seeded = false;
  }

  // Construct a SEEDED poller from a pre-lowered `DefinitionSnapshot` (S1's `lowerDefinitions`) —
  // the static-definitions self-host path (S2). Same `DefinitionPoller` type, so `local.poller`'s
  // type is unchanged and the adapter's resolve/lifecycle path is byte-for-byte unchanged. It carries
  // NO definitions endpoint / privileged credential / fetch — structurally unable to reach a remote
  // source. `isReady()` is true from construction (seed loaded, non-empty); `start()`/`stop()` are
  // real no-ops (no thread, no timer, no URL, no fetch). The private constructor above stays the
  // fetching mode; this factory is the only seeded entry.
  static seeded(snapshot: DefinitionSnapshot): DefinitionPoller {
    const poller = Object.create(DefinitionPoller.prototype) as DefinitionPoller;
    Object.assign(poller, {
      definitionsUrl: undefined,
      definitionsKey: undefined,
      pollIntervalMs: undefined,
      doFetch: undefined,
      snapshot,
      loadedSuccessfullyOnce: true,
      timer: undefined,
      loadingPromise: undefined,
      stopped: false,
      seeded: true,
    });
    return poller;
  }

  // Kick off polling: an immediate first load, then a self-rescheduling interval. Returns the
  // first-load promise so a caller can await readiness. In seeded mode this is a REAL no-op — the
  // seed is already present, so there is no thread, no URL resolution, and no fetch (the adapter's
  // constructor calls `start()` unconditionally; a seeded poller that fetched would reintroduce the
  // very remote dependency S2 bypasses).
  start(): Promise<void> {
    if (this.seeded) {
      return Promise.resolve();
    }
    return this.load();
  }

  // True once at least one successful load has parsed a non-empty definition list — S2's
  // "should I try local eval" gate.
  isReady(): boolean {
    return this.loadedSuccessfullyOnce && this.snapshot.flags.length > 0;
  }

  // The current parsed definition snapshot, read atomically by the evaluator. Before the first
  // successful load this is the frozen empty snapshot (never undefined) so a read never crashes.
  getSnapshot(): DefinitionSnapshot {
    return this.snapshot;
  }

  // Halt polling: clear the pending timer so no further loads are scheduled. Idempotent.
  stop(): void {
    this.stopped = true;
    if (this.timer !== undefined) {
      clearTimeout(this.timer);
      this.timer = undefined;
    }
  }

  // Load definitions once, deduping a concurrent call onto the same in-flight promise. Reschedules
  // the next poll BEFORE doing the fetch work (mirroring the reference), so a slow/failed fetch
  // never stalls the interval.
  private load(): Promise<void> {
    // A seeded poller never fetches (its `start()` short-circuits); this guard makes that structural
    // even against a direct `load()` — the fetch machinery is undefined, so there is nothing to load.
    if (this.seeded) {
      return Promise.resolve();
    }
    if (this.loadingPromise !== undefined) {
      return this.loadingPromise;
    }
    this.scheduleNext();
    this.loadingPromise = this.fetchDefinitions()
      .catch(() => {
        // A failed load leaves the prior snapshot in place — never overwrite good data with an
        // error. The next scheduled poll retries.
      })
      .finally(() => {
        this.loadingPromise = undefined;
      });
    return this.loadingPromise;
  }

  private scheduleNext(): void {
    if (this.stopped) {
      return;
    }
    if (this.timer !== undefined) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      void this.load();
    }, this.pollIntervalMs);
    // Node's timer handle carries .unref() — call it so a short-lived process (CLI/cron/test) that
    // builds a local-capable flag client isn't held open up to the poll interval by the pending
    // reschedule (mirrors batch-queue's armInterval; browser/edge setTimeout returns a bare number,
    // where the optional-chain makes this a no-op).
    this.timer?.unref?.();
  }

  private async fetchDefinitions(): Promise<void> {
    // Narrows the optional fetch machinery: it is defined only in fetching mode, and `load()` (the
    // only caller) already short-circuits in seeded mode — so this is unreachable when absent.
    if (
      this.doFetch === undefined ||
      this.definitionsUrl === undefined ||
      this.definitionsKey === undefined
    ) {
      return;
    }
    const response = await this.doFetch(this.definitionsUrl, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.definitionsKey}`,
      },
    });
    if (response.ok === false) {
      return;
    }
    const parsed = (await response.json()) as DefinitionsWireResponse;
    const flags = parsed[FLAGS_WIRE_KEY] ?? [];
    this.snapshot = Object.freeze({
      flags,
      flagsByKey: Object.freeze(
        flags.reduce<Record<string, FlagDefinition>>((acc, flag) => {
          acc[flag.key] = flag;
          return acc;
        }, {})
      ),
      groupTypeMapping: Object.freeze({ ...(parsed[GROUP_TYPE_MAPPING_WIRE_KEY] ?? {}) }),
      cohorts: Object.freeze({ ...(parsed[COHORTS_WIRE_KEY] ?? {}) }),
    });
    this.loadedSuccessfullyOnce = true;
  }
}

// Resolve the definitions URL from the consumer's bare flag origin: append the [WIRE] definitions
// path and the token + send-cohorts query params.
function resolveDefinitionsUrl(definitionsEndpoint: string, token: string): string {
  const params = new URLSearchParams({ [TOKEN_WIRE_QUERY]: token });
  params.set(SEND_COHORTS_WIRE_QUERY, '');
  return `${joinHostPath(definitionsEndpoint, DEFINITIONS_ENDPOINT_WIRE_PATH)}?${params.toString()}`;
}
